File: app/src/main/java/com/example/spelltracker/Spell.java
package com.example.spelltracker;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

/**
 * Модель заклинания D&D.
 *
 * <p>Является сущностью Room (таблица {@code spell}). При первом запуске
 * приложения данные импортируются из CSV в локальную БД, после чего
 * заклинания читаются исключительно из БД — это позволяет избежать
 * {@code TransactionTooLargeException} при передаче большого списка
 * заклинаний (1+ МБ) в аргументах фрагмента через {@link android.os.Parcel}.</p>
 */
@Entity(
        tableName = "spell",
        indices = {
                @Index("level"),
                @Index(value = "name")
        }
)
public class Spell {

    @PrimaryKey(autoGenerate = true)
    public long id;

    @NonNull
    @ColumnInfo(name = "name")
    public String name = "";

    @ColumnInfo(name = "school")
    public String school;

    @ColumnInfo(name = "level")
    public int level;

    @ColumnInfo(name = "casting_time")
    public String castingTime;

    @ColumnInfo(name = "range_text")
    public String range;

    @ColumnInfo(name = "components")
    public String components;

    @ColumnInfo(name = "duration")
    public String duration;

    @ColumnInfo(name = "description")
    public String description;

    @ColumnInfo(name = "higher_level")
    public String higherLevel;

    /** Пустой конструктор для Room. */
    public Spell() {
    }

    public Spell(@NonNull String name, String school, int level,
                 String castingTime, String range, String components,
                 String duration, String description, String higherLevel) {
        this.name = name;
        this.school = school;
        this.level = level;
        this.castingTime = castingTime;
        this.range = range;
        this.components = components;
        this.duration = duration;
        this.description = description;
        this.higherLevel = higherLevel;
    }

    public String levelLabel() {
        return level == 0 ? "Заговор" : level + " ур.";
    }
}

File: app/src/main/java/com/example/spelltracker/data/Spell.kt
package com.example.spelltracker.data

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * Заклинание D&D 5e — нормализованная запись (version 4).
 *
 * Поля один-к-одному соответствуют ключам в
 * `app/build/generated/assets/spells_normalized.json`, которое генерируется
 * build-time Gradle-таской `generateSpellsDb`. Один wide-table подход:
 * справочник всего ~800 строк, джойны не нужны.
 *
 * Multi-value поля (subclasses / races / classes / savingThrows) — CSV-строки.
 * Для поиска «содержит X» достаточно `csv.contains("X")`.
 *
 * Эволюция схемы:
 *   v3 (предыдущая) — 11 полей, данные из `assets/spells_normalized.json`
 *   (при старом pipeline'е: CSV + class_*.json, id = name.hashCode()).
 *   v4 (текущая) — 24 поля, данные из `assets/spells_normalized.json`,
 *     id = стабильный numeric из source-системы. Миграция destructive
 *     через `fallbackToDestructiveMigration` — справочник приходит из
 *     assets, потеря не страшна. См. Этап N в QWEN.md.
 */
@Entity(tableName = "spells")
data class Spell(
    @PrimaryKey val id: Long,
    val name: String,
    val nameEng: String,
    val source: String,
    val sourceGroup: String,
    val level: Int,
    val school: String,
    val ritual: Boolean,
    val concentration: Boolean,
    val timecast: String,
    val distance: String,
    val duration: String,
    val componentV: Boolean,
    val componentS: Boolean,
    val componentM: Boolean,
    val materialConsumed: Boolean,
    val materialDesc: String,
    val descriptionHtml: String,
    val upperLevel: String,
    val url: String,
    val classes: String,
    /** Имена подклассов, CSV. Парный список к [subclassParents]. */
    val subclasses: String,
    /**
     * English id класса-родителя для каждого подкласса в [subclasses], CSV.
     * Нужен runtime-фильтру «показать подклассы выбранных классов».
     * Пример: subclasses="Домен Войны,Клятва Мести", subclassParents="cleric,paladin".
     */
    val subclassParents: String,
    val races: String,
    val savingThrows: String,
)
